frame/contracts/src/storage/meter.rs

```diff
@@ -329,12 +332,24 @@ where
 	///
 	/// This tries to [`Ext::check_limit`] on `origin` and fails if this is not possible.
 	pub fn new(
-		origin: &T::AccountId,
+		origin: &Caller<T>,
 		limit: Option<BalanceOf<T>>,
 		min_leftover: BalanceOf<T>,
 	) -> Result<Self, DispatchError> {
-		let limit = E::check_limit(origin, limit, min_leftover)?;
-		Ok(Self { limit, ..Default::default() })
+		// Check the limit only if the origin is not root.
+		return match origin {
+			Caller::Root => {
+				match limit {
+					// If the limit is specified, that the root's limit.
+					Some(l) => Ok(Self { limit: l, ..Default::default() }),
+					None => Err(<Error<T>>::StorageDepositLimitRequired.into()),
```

Comment:
why?

Reply:
FYI, we're introducing a fallback limit for storage deposit being in https://github.com/paritytech/substrate/pull/13565
It's being enforced in case no limit was specified for the call. 

Meanwhile, for a call being initiated by *Root*, we probably want the [_charging_](https://github.com/paritytech/substrate/blob/d36d70e7b7b5910acab634e5f03cb88aca73a5a6/frame/contracts/src/storage/meter.rs#L498) part of the storage meter to behave differently.

Reply:
A fallback limit for storage sounds great, it will help with this.

Regarding the [charge](https://github.com/paritytech/substrate/blob/d36d70e7b7b5910acab634e5f03cb88aca73a5a6/frame/contracts/src/storage/meter.rs#L498) fn, it doesn't accept a `Origin`/`Caller` origin, it only accepts an `AccountId` as an input. This way Root never reaches that fn, you can see it within [into_deposit](https://github.com/paritytech/substrate/blob/d36d70e7b7b5910acab634e5f03cb88aca73a5a6/frame/contracts/src/storage/meter.rs#L364)